woocommerce/woocommerce
# Refactor the settings pages classes

**Note:** This pull request has been superseded by [a newer one](https://github.com/woocommerce/woocommerce/pull/27684) due to the difficulty of merging the changes from master.

This pull request introduces an extensive refactor of the WooCommerce settings classes (`WC_Settings_Page` and derived classes) that includes the following:

* Modify the `get_sections` method so that it invokes a new protected `get_own_sections` method and then triggers the `'woocommerce_get_sections_' . id` filter.

  * This way the filter is triggered only in the base class and not in each of the derived classes too.
<br/>

* Change the `get_settings()` method so that it has its signature changed to `get_settings( $current_section = '' )` in the base class and in all the derived class.

  * Some derived classes were already using this signature, but others (those not having multiple sections natively) weren't, making then effectively impossible to define multiple sections for these pages via filters.
<br/>

* Change the implementation of `get_settings` in the base class so that it searches for a `get_settings_for_{section_id}_section` method in the class and executes it, otherwise it executes the new protected method `get_settings_for_section( $current_section )`; then it triggers the `'woocommerce_get_settings_' . id` filter.

  * This makes it easier to separate the code that returns the list of filters in multiple methods, one per section, instead of using one big if-else-else... block.
  * Also, `'woocommerce_get_settings_' . id` is triggered in one single place too.
<br/>

* General refactorings to fix code sniffer errors, including replacing old-style sniffer directives with phpcs:enable/disable/ignore comments.

* Some duplicated code has been refactored to protected methods.

* Changes the "todo: remove in 4.0" comment in `WC_Settings_Rest_API` to "remove in 5.0".

**TODO:** Add unit tests.


### How to manually test the changes in this Pull Request:

Create the following plugin and open the _General_ settings tab. You'll see that it has three sections: `General, Foobar, Fizzbuz`. The first one has one extra setting, the second one has just one setting, and the thirs one is empty. Without the changes introduced in this pull request you would have been able to create the sections, but the `woocommerce_get_settings_general` trigger would be invoked with only the `$settings` parameter and thus you wouldn't have been able to return different per section settings.

```
<?php
/*
Plugin Name: Foobar & Fizzbuzz
Version: 1.0
Description: Foo? Bar. Fizz? Buzz. See? It's easy!
Author: WooCommerce (testing)
*/

add_filter('woocommerce_get_sections_general', 'foobar_sections_for_general_settings', 10, 1);

function foobar_sections_for_general_settings($settings) {
	$settings['foobars'] = 'Foobars';
	$settings['fizzbuzzs'] = 'Fizzbuzzs';
	return $settings;
}

add_filter('woocommerce_get_settings_general', 'foobar_settings_for_general_tab', 10, 2);

function foobar_settings_for_general_tab($settings, $current_section) {
	if( '' === $current_section ) {
		$settings[] =
			array(
				'title' => "Bonus settings",
				'type'  => 'title',
				'desc'  => "Some bonus settings inside the default section!"
			);

		$settings[] =
			array(
				'title'    => 'Some setting',
				'desc'     => 'Is this setting set or not?',
				'id'       => 'foobar_some_setting',
				'default'  => 'no',
				'type'     => 'checkbox',
				'desc_tip' => 'A setting that can be set or not.'
			);

		$settings[] = array( 'type' => 'sectionend'	);

		return $settings;
	}
	else if( 'foobars' === $current_section ) {
		return array(
			array(
				'title' => "Foobars",
				'type'  => 'title',
				'desc'  => "What about foobars?"
			),

			array(
				'title'    => 'The foo is bar',
				'desc'     => 'States if the foo is bar',
				'id'       => 'foobar_the_foo_is_bar',
				'default'  => 'no',
				'type'     => 'checkbox',
				'desc_tip' => 'Is the foo bar enough?',
			),

			array( 'type' => 'sectionend' )
		);
	}
	else {
		return array();
	}
}
```

### All Submissions:






### Changes proposed in this Pull Request:



Closes # .



### Other information:




### Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.
